File: app/src/main/java/com/shindra/Line/LineActivity.java
package com.shindra.Line;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.shindra.Misc.ErrorDialog;
import com.shindra.Misc.LoadingDialog;
import com.shindra.Misc.MyViewModel;
import com.shindra.R;
import com.shindra.arrakis.observable.ObservableExtensionKt;
import com.shindra.arrakis.observable.ObservableListener;
import com.shindra.ctslibrary.apibo.RouteType;
import com.shindra.ctslibrary.bo.Line;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class LineActivity extends AppCompatActivity
{
    public LoadingDialog loadingDialog;
    public ErrorDialog errorDialog;
    public LineFragment fragment;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_line);

        loadingDialog = new LoadingDialog(this);
        errorDialog = new ErrorDialog(this);

        fragment = new LineFragment();
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.add(R.id.frameContainer, fragment).commit();

        MyViewModel model = new ViewModelProvider(this).get(MyViewModel.class);
        ObservableExtensionKt.observe(model.lines(), new ObservableListener<ArrayList<Line>>()
        {
            @Override
            public void onLoading()
            {
                //call once we started the network called. Indicate that the network call is in progress
                //Show the loading dialog
                loadingDialog.show();
            }

            @Override
            public void onSuccess(ArrayList<Line> data)
            {
                //call once the network call has responded with a success
                //Set the right title to the app
                getSupportActionBar().setTitle(R.string.lines);

                //Dismiss the loading dialog
                loadingDialog.dismiss();

                //Only keep Tram lines
                ArrayList<Line> tramLines = new ArrayList<Line>();
                for (Line item : data)
                    if (item.getRouteType() == RouteType.TRAM)
                        tramLines.add(item);

                //Update the recycler view through the adapter
                fragment.updateWidgets(tramLines);
            }

            @Override
            public void onError(@NotNull Throwable throwable)
            {
                //call if the network call has responded with an error
                //Dismiss the loading dialog and show the error dialog
                loadingDialog.dismiss();
                errorDialog.show();
            }
        });
    }
}

File: app/src/main/java/com/shindra/LineActivity.kt
package com.shindra

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.shindra.arrakis.observable.ObservableListener
import com.shindra.arrakis.observable.observe
import com.shindra.ctslibrary.apibo.RouteType
import com.shindra.ctslibrary.bo.Line

class LineActivity : AppCompatActivity(), LineViewHolder.OnLineClickListener {

    private var tramList = ArrayList<Line>()

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.line_activity)
        setTitle(R.string.line_activity_name)

        val tramRecyclerList = findViewById<RecyclerView>(R.id.line_list)
        tramRecyclerList.layoutManager = LinearLayoutManager(this)
        tramRecyclerList.adapter = LineAdapter(tramList, this)

        val loadingDialog = LoadingDialog(this)

        val model = ViewModelProvider(this).get(MyViewModel::class.java)
        model.lines().observe(object : ObservableListener<ArrayList<Line>> {

            override fun onLoading() {
                //call once we started the network called. Indicate that the network call is in progress
                loadingDialog.showDialog()
            }

            override fun onSuccess(data: ArrayList<Line>) {
                //call once the network call has responded with a success
                for (line in data) {
                    if (line.routeType == RouteType.TRAM) {
                        tramList.add(line)
                    }
                }
                (tramRecyclerList.adapter as LineAdapter).lines = tramList
                (tramRecyclerList.adapter as LineAdapter).notifyDataSetChanged()
                loadingDialog.dismissDialog()
            }

            override fun onError(throwable: Throwable) {
                //call if the network call has responded with an error
            }
        })
    }

    override fun onItemClick(line: Line) {
        val intent = Intent(this, ScheduleActivity::class.java)
        intent.putExtra("lineName", line.name)
        startActivity(intent)
    }
}